### ISSUE TITLE
TokenCache typing problem

### ISSUE THREAD

**Issue by bzums** [zalando-incubator/authmosphere#170]
If initialising the `TokenCache` like:

```ts
const oAuthConfig = {
  credentialsDir,
  accessTokenEndpoint,
  tokenInfoEndpoint,
  grantType: OAuthGrantType.CLIENT_CREDENTIALS_GRANT
};

const cache = new TokenCache(tokenConfig, oAuthConfig);
```

we currently get the following error:

```
> src/services/token-cache.ts(29,38): error TS2345: Argument of type '{ credentialsDir: string; accessTokenEndpoint: string; tokenInfoEndpoint: string; grantType: OAuthGrantType; }' is not assignable to parameter of type 'TokenCacheOAuthConfig'.
>   Type '{ credentialsDir: string; accessTokenEndpoint: string; tokenInfoEndpoint: string; grantType: OAuthGrantType; }' is not assignable to type 'CredentialsClientConfig & GrantConfigBase & { grantType: OAuthGrantType.REFRESH_TOKEN_GRANT; refreshToken: string; } & { tokenInfoEndpoint: string; }'.
>     Type '{ credentialsDir: string; accessTokenEndpoint: string; tokenInfoEndpoint: string; grantType: OAuthGrantType; }' is not assignable to type 'CredentialsClientConfig'.
>       Property 'clientId' is missing in type '{ credentialsDir: string; accessTokenEndpoint: string; tokenInfoEndpoint: string; grantType: OAuthGrantType; }'.
```

which can be fixed via an explicitly typing:

```ts
const oAuthConfig: TokenCacheOAuthConfig = { ... };
```

However, that should not be necessary for the user. We have to investigate further why this problem occurs.

**Comment by ISO50** [comment]
@bzums 

When using the above method Typescript infers `OAuthGrantType` instead of `OAuthGrantType.CLIENT_CREDENTIALS_GRANT` for the `grantType` field.

This can be fixed by:

1. `oAuthConfig: TokenCacheOAuthConfig`
2. inlining: 

```ts
const cache = new TokenCache(tokenConfig, {
  credentialsDir: 'credentialsDir',
  accessTokenEndpoint: 'accessTokenEndpoint',
  tokenInfoEndpoint: 'tokenInfoEndpoint',
  grantType: OAuthGrantType.CLIENT_CREDENTIALS_GRANT
});
```

3. Or using Type assertions:

```ts
const oAuthConfig = {
  credentialsDir: 'credentialsDir',
  accessTokenEndpoint: 'accessTokenEndpoint',
  tokenInfoEndpoint: 'tokenInfoEndpoint',
  grantType: OAuthGrantType.CLIENT_CREDENTIALS_GRANT as OAuthGrantType.CLIENT_CREDENTIALS_GRANT,
};
```

--- 

Maybe using String Literals instead of Enums could make that work without this workarounds.

**Comment by ISO50** [comment]
One of the underlying problems, read here: https://github.com/Microsoft/TypeScript/issues/9489

**Comment by ISO50** [comment]
https://stackoverflow.com/questions/36415722/typescript-string-literal-with-duck-typed-object

**Comment by ISO50** [comment]
Fixed by #171
